SICP_book/chapter_1_2.metta

```diff
@@ -0,0 +1,645 @@
+        (if (even? $exp)
+            (remainder (sqr (expmod $base (/ $exp 2) $m)) $m)
+            (remainder (* $base (expmod $base (dec $exp) $m)) $m))))
+
+; randomint! is from additional_funcs.py extension
+(= (random $end) (randomint! 0 $end))
+
+(= (ferma-test $n)
+    (let $try-it
+        (lambda1 $a
+            (== (expmod $a $n $n) $a))
+            ($try-it (inc (random (dec $n))))))
+
+(= (fast-prime? $n $times)
+    (if (== $times 0)
+        True
+        (if (ferma-test $n)
+            (fast-prime? $n (dec $times))
+            False)))
+
+; These two asserts could possibly fail due to non-deterministic nature of fast-prime? function.
+!(assertEqual
+    (fast-prime? 17 5)
+    True)
+
+!(assertEqual
+    (fast-prime? 16 5)
+    False)
+
+; Exercise 1.21.  Use the smallest-divisor procedure to find the
+; smallest divisor of each of the following numbers: 199, 1999, 19999.
+
+!(assertEqual
+    (smallest-divisor 199)
+  199)
+
+; These two takes too much time to evaluate so I've commented these lines.
+;!(smallest-divisor 1999)
+;!(smallest-divisor 19999)
+
+
+; Exercise 1.22.
+;
+; Most Lisp implementations include a primitive called runtime that returns an
+; integer that specifies the amount of time the system has been running
+; (measured, for example, in microseconds). The following timed-prime-test procedure,
+; when called with an integer n, prints n and checks to see if n is prime.
+;
+; If n is prime, the procedure prints three asterisks followed by the amount of time used in performing the test.
+
+(= (timed-prime-test $n)
+  (let* (
+    (() (println! $n))
+    (() (let $curtime (timems!) (start-prime-test $n $curtime))))
+    () ))
+
+(= (start-prime-test $n $start-time)
+  (if (prime? $n)
+      (report-prime (- (timems!) $start-time))
+      (println! "not simple")))
+
+(= (report-prime $elapsed-time)
+    (let*
+    (
+        (() (println! " time in ms: "))
+        (() (println! $elapsed-time))
+    )
+    ()))
+
+; Using this procedure, write a procedure search-for-primes that checks the primality of
+; consecutive odd integers in a specified range.
+; Use your procedure to find the three smallest primes larger than: 1000, 10,000, 100,000 and 1,000,000.
+;
+; Note the time needed to test each prime.
+
+; This function is right but it works longer than not right one.
+(= (loop-search-for-primes $start $end)
+    (if (>= $start $end)
+        (println! "done searching")
+        (if (even? $start)
+            (loop-search-for-primes (inc $start) $end)
+            (let*
+            (
+                (() (timed-prime-test $start))
+                (() (loop-search-for-primes (inc $start) $end))
+            )
+            ()))))
+
+; !(loop-search-for-primes 10 18)
+; Output primes: 11 13 17, time on each prime: 63, 114, 401 ms. Time is only evaluated for checking if number is a prime
+; not for entire loop-search function. Entire search is much longer. Strange thing is that if we will call timed-prime-test
```

Comment:
Not too strange taking into account two items:
- you are measuring times only once, thus 44 and 46 has some difference, if you run experiment thousands of times you would see this numbers deviate rarely
- test for 17 has one more step because ceil of square root of 17 is 4 while for 11 and 13 it is 3, thus in this case the prime checking procedure always makes one additional step